Uncouple emptying gas_meter from updating the gas spent for the block

aliXsed opened the pull request:
Closes #67

Commit:
Uncouple emptying gas_meter from updating the gas spent for the block. However keep the update function for GasHandler as it seems to be relevant to this trait.

aliXsed commented on frame/contracts/src/gas.rs:
```diff
@@ -209,6 +209,18 @@ pub trait GasHandler<T: Trait> {
+		// also has Gas type.
+		GasSpent::mutate(|block_gas_spent| *block_gas_spent += gas_meter.spent());
+
+		Self::empty_unused_gas(transactor, gas_meter);
```
I'm uncertain whether this would actually call the overridden function in the places that we have implemented this trait. I will do a test soon.

aliXsed replied:
My test showed this should work. So we can go ahead.

jordy25519 replied:
Can you add a specific unit test to see a custom impl updates gas_spent

aliXsed replied:
I made a change so customized_fill_gas_does_not_charge_the_user, if passed, would prove the right empty_unused_gas is called.

cowboy-bebug commented on frame/contracts/src/gas.rs:
```diff
@@ -257,14 +269,8 @@ pub fn refund_unused_gas<T: Trait>(
-	// This cannot overflow, since `gas_spent` is never greater than `block_gas_limit`, which
-	// also has Gas type.
-	GasSpent::mutate(|block_gas_spent| *block_gas_spent += gas_spent);
-
```
Do we need to call the update method here to ensure the default behaviour here (please ignore me if we don't need to)?
```rust
Self::update(transactor, gas_meter);
```

aliXsed replied:
It's now the other way around. The update method's default is to call this function. So the default behaviour would remain as it was before.

cowboy-bebug commented:
~I think the build errors are coming from the way we use the latest nightly in our circleci config. Let me fix it in another PR~ Actually the latest nightly probably fixes this.

jordy25519 commented:
related: https://github.com/rustwasm/wasm-bindgen/issues/2009

jordy25519 commented on frame/contracts/src/gas.rs:
```diff
@@ -209,6 +209,18 @@ pub trait GasHandler<T: Trait> {
+	/// This function should be called when the contract has stopped consuming gas and the gas_meter
+	/// is ready to be read. This will update the gas spent for the block and then empties the
+	/// unused gas.
+	fn update(transactor: &T::AccountId, gas_meter: GasMeter<T>) {
```
a more specific name would be good here

aliXsed replied:
Done.

Commit:
Choose a better name for update function. Refine a test that would show "finish" will call the right "empty_unused_gas".

Commit:
Test gas handler updates the gas spent for the block on finish

hoani commented on frame/contracts/src/gas.rs:
```diff
@@ -209,6 +209,18 @@ pub trait GasHandler<T: Trait> {
+	/// This function should be called when the contract has stopped consuming gas and the gas_meter
+	/// is ready to be read. This will update the gas spent for the block and then empties the
+	/// unused gas.
+	fn finish(transactor: &T::AccountId, gas_meter: GasMeter<T>) {
```
I don't think that `finish` is more descriptive than `update`, but I will let @holygits decide since it was his concern in the first place.

hoani commented on frame/contracts/src/gas.rs:
```diff
@@ -209,6 +209,18 @@ pub trait GasHandler<T: Trait> {
+		// Increase total spent gas.
+		// This cannot overflow, since `gas_spent` is never greater than `block_gas_limit`, which
+		// also has Gas type.
+		GasSpent::mutate(|block_gas_spent| *block_gas_spent += gas_meter.spent());
```
From my understanding, we want to separate contract logic from gas handler logic.
The reason is so that anyone who implements a gas handler can choose how to pay for gas, but we don't want that user to determine how a contract tracks gas spent.
If we put `finish` in the `GasHandler`, then the user still decides how the contract tracks gas.

In my opinion, to follow these rules, we should insert:
`GasSpent::mutate(|block_gas_spent| *block_gas_spent += gas_meter.spent());`
On line 592 (https://github.com/plugblockchain/plug-blockchain/blob/613e545f8ac6f07abdf8eacd7457c18fece6f0af/frame/contracts/src/lib.rs#L592)
and line 760 (https://github.com/plugblockchain/plug-blockchain/blob/613e545f8ac6f07abdf8eacd7457c18fece6f0af/frame/contracts/src/lib.rs#L760)
of the Contracts `lib.rs` so that full control is under the contract for how it tracks gas spent

jordy25519 replied:
We could do some kind of decorator but this would be so much simpler XD

aliXsed replied:
The logic that updates the gas spent for the block is still relevant to the gas handler. The fact is a gas meter has two data. The gas spent that should be used to update the gas spent for the block and the gas left which is used in emptying unused gas. It seemed relevant to me to bundle these two into a function that would be used before finishing a contract as we would never want for one of these two to happen without the other one. 
On the other hand, the real aim was, we wanted to make sure if "empty_unused_gas" is overridden, the gas spent for the block would still keep being updated without exposing any new gas specific function of Plug to other projects like cennznet.

hoani reviewed (changes requested):
Please see requested changes, happy to discuss design decisions =)

Commit:
In lining the call for mutating the gas-spent for blocks. This is because the team see a new interface such as finish fo GasHandler as exposing.

The test for checking if gas-spent for the block increases correctly is removed as it couldn't be relevant with this new design.

jordy25519 reviewed (approved):
TY LGTM

hoani reviewed (approved):
LGTM 👍

cowboy-bebug reviewed (approved):
LGTM !
